include/phasar/PhasarLLVM/Utils/AnalysisPrinter.h

```diff
@@ -0,0 +1,67 @@
+#ifndef PHASAR_PHASARLLVM_UTILS_ANALYSISPRINTER_H
+#define PHASAR_PHASARLLVM_UTILS_ANALYSISPRINTER_H
+
+#include "phasar/Domain/BinaryDomain.h"
+#include "phasar/PhasarLLVM/Utils/LLVMShorthands.h"
+
+#include <optional>
+#include <type_traits>
+#include <vector>
+
+namespace psr {
+
+template <typename AnalysisDomainTy> struct Warnings {
+  typename AnalysisDomainTy::n_t Instr;
+  typename AnalysisDomainTy::d_t Fact;
+  typename AnalysisDomainTy::l_t LatticeElement;
+
+  // Constructor
+  Warnings(typename AnalysisDomainTy::n_t Inst,
+           typename AnalysisDomainTy::d_t DfFact,
+           typename AnalysisDomainTy::l_t Lattice)
+      : Instr(std::move(Inst)), Fact(std::move(DfFact)),
+        LatticeElement(std::move(Lattice)) {}
+};
+
+template <typename AnalysisDomainTy> struct Results {
+  std::vector<Warnings<AnalysisDomainTy>> War;
+};
+
+// TODO: explicit-template-instantiation
+template <typename AnalysisDomainTy> class AnalysisPrinter {
+public:
+  virtual ~AnalysisPrinter() = default;
+  AnalysisPrinter() : AnalysisResults{.War = {}} {}
+
+  AnalysisPrinter(const AnalysisPrinter &) = delete;
+  AnalysisPrinter &operator=(const AnalysisPrinter &) = delete;
+
+  AnalysisPrinter(AnalysisPrinter &&) = delete;
+  AnalysisPrinter &operator=(AnalysisPrinter &&) = delete;
+
+  virtual void onResult(Warnings<AnalysisDomainTy> War) {
+    AnalysisResults.War.emplace_back(std::move(War));
+  }
+
+  virtual void onInitialize(){};
+  virtual void onFinalize(llvm::raw_ostream &OS = llvm::outs()) const {
+    for (auto Iter : AnalysisResults.War) {
+
+      OS << "\nAt IR statement: " << psr::nToString(Iter.Instr) << "\n";
+
+      OS << "\tFact: " << psr::dToString(Iter.Fact) << "\n";
+
+      if constexpr (std::is_same_v<typename AnalysisDomainTy::l_t,
+                                   BinaryDomain>) {
+        OS << "Value: " << psr::lToString(Iter.LatticeElement) << "\n";
+      }
+    }
+  }
+
+private:
```

Comment:
What if, I want to create a derived class of AnalysisPrinter without class variable. In the current version, we will still allocate memory to the class variable and not use it (similar to NullAnalysisPrinter). To overcome this, you could introduce another super class follows,

```
template <typename AnalysisDomainTy> class AnalysisPrinterBase {
  virtual void onResult(Warnings<AnalysisDomainTy> War) = 0;
  virtual void onInitialize() = 0;
  virtual void onFinalize(llvm::raw_ostream &OS = llvm::outs()) = 0;
}
```

and rename current AnalysisPrinter to DefaultAnalysisPrinter and inherit it from AnalysisPrinterBase as follows,

```
template <typename AnalysisDomainTy> class DefaultAnalysisPrinter: public AnalysisPrinter {
....
}
```

and so, NullAnalysisPrinter would be as follows,

```
template <typename AnalysisDomainTy>
class NullAnalysisPrinter final : public AnalysisPrinterBase<AnalysisDomainTy> {
...
}
```